assets/js/base/utils/errors.js

```diff
@@ -1,22 +1,30 @@
-const formatErrorMessage = ( error, messageProperty = 'frontendMessage' ) => {
-	if ( typeof error === 'object' && error.hasOwnProperty( 'message' ) ) {
-		return {
-			[ messageProperty ]: error.message,
-		};
-	}
-
-	return error;
-};
-
+/**
+ * Given a JS error or a fetch response error, parse and format it so it can be displayed to the user.
+ *
+ * @param {object} error Error object.
+ * @param {function} [error.json] If a json method is specified, it will try parsing the error first.
+ * @param {string} [error.message] If a message is specified, it will be shown to the user.
+ * @param {string} [error.type] The type will be used to determine how the error is shown to the user.
+ * @return {object} Error object containing a message and type.
+ */
 export const formatError = async ( error ) => {
-	if ( error.json ) {
+	if ( typeof error.json === 'function' ) {
 		try {
 			const parsedError = await error.json();
-			return formatErrorMessage( parsedError, 'apiMessage' );
+			return {
+				message: parsedError.message,
+				type: parsedError.type || 'api',
+			};
 		} catch ( e ) {
-			return formatErrorMessage( e );
+			return {
+				message: e.message,
+				type: 'frontend',
+			};
 		}
 	}
 
-	return formatErrorMessage( error );
+	return {
+		message: error.message,
+		type: error.type || 'frontend',
+	};
```

Comment:
I like the direction with providing a "type" to the error.  However:

- defaults vary in the logic.  Why?
- any caught exception _always_ defaults to `frontend` as the type. Why? 

I'm not sure we've landed on a good api for error handling yet, but it probably should be iterated on in it's own pull.

Reply:
The rationale behind that was that an API error will be a `Response` object, so it will have a `json` method, while errors produced in the frontend will not have that method.

Reply:
Right, there's a bit of code smell here with the confusing logic.  Let's iterate on this in a follow-up (not urgent).